File: scribe/analyzer.py
"""Meeting analysis with AssemblyAI LeMUR.

LeMUR receives the transcript plus a prompt asking for a strict JSON
payload: meeting metadata (title, summary, key points, decisions, action
items) and a list of knowledge topics with distilled Markdown content.

The prompt includes the topics already present in the library so the model
reuses existing slugs when the subject matches and only creates new slugs
for genuinely new subjects.

``build_prompt`` and ``parse_analysis`` are shared with the Kimi backend
(``kimi_analyzer.py``), which produces the same JSON contract through a
local Kimi Code CLI subprocess instead of LeMUR.

Parsing is defensive: the JSON block is extracted even when wrapped in
prose or ``` fences; on total failure we fall back to a minimal result with
a single "general" topic holding the summary.
"""

from __future__ import annotations

import json
import logging
import re
from dataclasses import dataclass, field

log = logging.getLogger(__name__)

_JSON_SCHEMA_EXAMPLE = """{
  "meeting": {
    "title": "short descriptive title",
    "summary": "2-4 sentence summary of the meeting",
    "key_points": ["point 1", "point 2"],
    "decisions": ["decision 1"],
    "action_items": [{"owner": "person name or 'unassigned'", "task": "what must be done"}]
  },
  "topics": [
    {
      "slug": "kebab-case-slug",
      "name": "Human readable name",
      "content": "Markdown with what was newly learned about this topic in this meeting"
    }
  ]
}"""


@dataclass
class Topic:
    slug: str
    name: str
    content: str


@dataclass
class AnalysisResult:
    title: str
    summary: str
    key_points: list[str] = field(default_factory=list)
    decisions: list[str] = field(default_factory=list)
    action_items: list[dict] = field(default_factory=list)  # [{"owner": ..., "task": ...}]
    topics: list[Topic] = field(default_factory=list)


def build_prompt(
    existing_topics: list[dict], language: str, transcript_path: str | None = None
) -> str:
    """Build the analyzer prompt. ``existing_topics`` is a list of {slug, name}.

    When ``transcript_path`` is given, the prompt instructs the model to read
    the transcript from that file (used by the Kimi backend, where embedding
    the whole transcript in the prompt would risk ARG_MAX on long meetings).
    """
    if existing_topics:
        topic_lines = "\n".join(f"- {t['slug']}: {t['name']}" for t in existing_topics)
        topics_block = (
            "The knowledge library already contains these topics (slug: name):\n"
            f"{topic_lines}\n"
            "REUSE an existing slug whenever the subject of a topic matches. "
            "Only create a new slug for genuinely new subjects."
        )
    else:
        topics_block = (
            "The knowledge library is empty yet. Create slugs for every subject discussed."
        )

    if transcript_path:
        source_block = (
            f"First, read the meeting transcript from this file using your file reading tool: "
            f"{transcript_path}\n"
            "Base your entire analysis on its contents."
        )
    else:
        source_block = "Analyze the meeting transcript provided with this request."

    language_rule = (
        f'Write ALL output (title, summary, key points, decisions, action items, topic names '
        f'and content) in the same language as the transcript ("{language}").'
        if language and language != "unknown"
        else "Write ALL output (title, summary, key points, decisions, action items, topic "
        "names and content) in the same language the transcript is written in."
    )

    return f"""You are analyzing a meeting transcript to distill it into a personal knowledge library.

{source_block}

{topics_block}

Respond with ONLY a valid JSON object (no prose, no markdown fences) matching exactly this schema:
{_JSON_SCHEMA_EXAMPLE}

Rules:
- Slugs must be lowercase kebab-case, ASCII, no accents.
- "content" of each topic is Markdown: what was newly learned about that topic in THIS meeting (facts, decisions, context), not a meeting summary.
- 1 to 5 topics, each focused on a single subject.
- {language_rule}
- If a list would be empty, use an empty array.
- Your final message must contain ONLY the JSON object, nothing else.
"""


def _extract_json(text: str) -> dict:
    """Extract the first JSON object from ``text``, tolerating fences/prose."""
    cleaned = text.strip()
    # Strip a wrapping ```json ... ``` fence if present.
    fence = re.search(r"```(?:json)?\s*(.*?)```", cleaned, re.DOTALL)
    if fence:
        cleaned = fence.group(1).strip()
    try:
        return json.loads(cleaned)
    except json.JSONDecodeError:
        pass
    # Fall back to the first balanced {...} block in the text.
    start = cleaned.find("{")
    end = cleaned.rfind("}")
    if start != -1 and end > start:
        return json.loads(cleaned[start : end + 1])
    raise ValueError("no JSON object found in analyzer response")


def _as_str_list(value) -> list[str]:
    if not isinstance(value, list):
        return []
    return [str(item) for item in value]


def _parse_response(payload: dict) -> AnalysisResult:
    meeting = payload.get("meeting")
    if not isinstance(meeting, dict):
        raise ValueError("missing 'meeting' object in LeMUR response")

    action_items = []
    for item in meeting.get("action_items") or []:
        if isinstance(item, dict):
            action_items.append(
                {"owner": str(item.get("owner", "unassigned")), "task": str(item.get("task", ""))}
            )
        elif item:
            action_items.append({"owner": "unassigned", "task": str(item)})

    topics = []
    for t in payload.get("topics") or []:
        if not isinstance(t, dict) or not t.get("slug"):
            continue
        topics.append(
            Topic(
                slug=str(t["slug"]),
                name=str(t.get("name") or t["slug"]),
                content=str(t.get("content") or ""),
            )
        )

    return AnalysisResult(
        title=str(meeting.get("title") or "Untitled meeting"),
        summary=str(meeting.get("summary") or ""),
        key_points=_as_str_list(meeting.get("key_points")),
        decisions=_as_str_list(meeting.get("decisions")),
        action_items=action_items,
        topics=topics,
    )


def parse_analysis(raw: str, backend_name: str = "analyzer") -> AnalysisResult:
    """Parse a raw analyzer text response into an AnalysisResult.

    Shared by every backend. Never raises on malformed input: falls back to a
    minimal result with a single "general" topic holding the raw response.
    """
    try:
        return _parse_response(_extract_json(raw))
    except (ValueError, KeyError, TypeError) as exc:
        log.warning(
            "%s response could not be parsed (%s); using minimal fallback",
            backend_name, exc,
        )
        summary = raw.strip()[:2000] or "Analysis failed to produce a parseable result."
        return AnalysisResult(
            title="Untitled meeting",
            summary=summary,
            topics=[Topic(slug="general", name="General", content=summary)],
        )


def analyze(transcript_result, existing_topics: list[dict], language: str) -> AnalysisResult:
    """Run a LeMUR task over a TranscriptResult and parse its JSON answer."""
    import assemblyai as aai

    sdk_transcript = transcript_result.sdk_transcript
    if sdk_transcript is None:
        raise RuntimeError("analyze() requires the SDK transcript object (missing in mock mode)")

    prompt = build_prompt(existing_topics, language)
    log.info("Running LeMUR analysis over transcript %s ...", transcript_result.id)
    response = sdk_transcript.lemur.task(prompt, final_model=aai.LemurModel.claude_sonnet_4_20250514)
    return parse_analysis(response.response or "", backend_name="LeMUR")

File: internal/analyzer/analyzer.go
// Package analyzer distills meeting transcripts into structured knowledge.
//
// Every backend answers the same prompt (BuildPrompt) with a strict JSON
// payload: meeting metadata (title, summary, key points, decisions, action
// items) and a list of knowledge topics with distilled Markdown content.
// The prompt includes the topics already present in the library so the
// model reuses existing slugs when the subject matches and only creates
// new slugs for genuinely new subjects.
//
// Parsing is defensive: the JSON block is extracted even when wrapped in
// prose or ``` fences; on total failure ParseAnalysis falls back to a
// minimal result with a single "general" topic holding the raw response.
//
// This is a port of scribe/analyzer.py (prompt builder and shared parser).
package analyzer

import (
	"embed"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/fernando143/patro/internal/logging"
	"github.com/fernando143/patro/internal/types"
)

// promptTemplate is the versioned prompt skeleton. It is authored and
// versioned in Langfuse Prompt Management; tools/promptsync fetches the
// "production"-labeled version and overwrites this file at release time
// (see .github/workflows/release.yml), so the built binary never talks to
// Langfuse itself. The {{source}}/{{topics}}/{{schema}}/{{language_rule}}
// placeholders are filled in by BuildPrompt below.
//
//go:embed prompt_template.txt
var promptTemplate string

// rawResponseSchema is the JSON schema example embedded verbatim in the
// prompt; the backends are instructed to match it exactly. Like
// promptTemplate, it is authored and versioned in Langfuse (as a separate
// prompt, so it can be validated on its own — see
// TestResponseSchemaMatchesParser) and synced by tools/promptsync.
//
//go:embed response_schema.json
var rawResponseSchema string

// responseSchema drops the file's trailing newline so substituting it into
// promptTemplate reproduces the exact prompt text the analyzer contract
// tests were pinned against.
var responseSchema = strings.TrimRight(rawResponseSchema, "\n")

// promptFragments holds the copy for BuildPrompt's conditional blocks —
// each is a separate Langfuse prompt (see tools/promptsync) because each
// condition (existing topics or not, transcript path or not, known
// language or not) needs its own independent wording, not one template
// with branches. Go still decides WHICH fragment applies; Langfuse only
// controls what each one says.
//
//go:embed prompt_fragments
var promptFragments embed.FS

// fragment reads and trims one file from promptFragments. Failure here
// means a fragment file is missing from the embedded build — a build-time
// defect, not a runtime condition, hence the panic.
func fragment(name string) string {
	data, err := promptFragments.ReadFile("prompt_fragments/" + name)
	if err != nil {
		panic("analyzer: missing embedded prompt fragment " + name + ": " + err.Error())
	}
	return strings.TrimRight(string(data), "\n")
}

var (
	topicsWithExistingFragment = fragment("topics_with_existing.txt")
	topicsEmptyFragment        = fragment("topics_empty.txt")
	sourceWithPathFragment     = fragment("source_with_path.txt")
	sourceWithoutPathFragment  = fragment("source_without_path.txt")
	languageKnownFragment      = fragment("language_known.txt")
	languageUnknownFragment    = fragment("language_unknown.txt")
)

// fencePattern matches a wrapping ```json ... ``` (or plain ``` ... ```)
// fence, capturing its body.
var fencePattern = regexp.MustCompile("(?s)```(?:json)?\\s*(.*?)```")

// BuildPrompt builds the analyzer prompt. existing holds {slug, name} refs
// of the topics already present in the library.
//
// When transcriptPath is given, the prompt instructs the model to read the
// transcript from that file (used by the CLI backends, where embedding the
// whole transcript in the prompt would risk ARG_MAX on long meetings).
//
// The output is byte-for-byte identical to scribe/analyzer.py's
// build_prompt.
func BuildPrompt(existing []types.TopicRef, language, transcriptPath string) string {
	topicsBlock := topicsEmptyFragment
	if len(existing) > 0 {
		lines := make([]string, len(existing))
		for i, t := range existing {
			lines[i] = fmt.Sprintf("- %s: %s", t.Slug, t.Name)
		}
		topicsBlock = strings.NewReplacer("{{topics_list}}", strings.Join(lines, "\n")).
			Replace(topicsWithExistingFragment)
	}

	sourceBlock := sourceWithoutPathFragment
	if transcriptPath != "" {
		sourceBlock = strings.NewReplacer("{{transcript_path}}", transcriptPath).
			Replace(sourceWithPathFragment)
	}

	languageRule := languageUnknownFragment
	if language != "" && language != "unknown" {
		languageRule = strings.NewReplacer("{{language}}", fmt.Sprintf("%q", language)).
			Replace(languageKnownFragment)
	}

	replacer := strings.NewReplacer(
		"{{source}}", sourceBlock,
		"{{topics}}", topicsBlock,
		"{{schema}}", responseSchema,
		"{{language_rule}}", languageRule,
	)
	return replacer.Replace(promptTemplate)
}

// ParseAnalysis parses a raw analyzer text response into an AnalysisResult.
//
// Shared by every backend. It never fails on malformed input: it logs a
// warning and falls back to a minimal result with a single "general" topic
// holding the raw response (trimmed to 2000 characters).
func ParseAnalysis(raw, backendName string) *types.AnalysisResult {
	payload, err := extractJSON(raw)
	if err == nil {
		var result *types.AnalysisResult
		if result, err = parseResponse(payload); err == nil {
			return result
		}
	}
	logging.Warnf("%s response could not be parsed (%s); using minimal fallback", backendName, err)

	summary := strings.TrimSpace(raw)
	if runes := []rune(summary); len(runes) > 2000 {
		summary = string(runes[:2000])
	}
	if summary == "" {
		summary = "Analysis failed to produce a parseable result."
	}
	return &types.AnalysisResult{
		Title:   "Untitled meeting",
		Summary: summary,
		Topics:  []types.Topic{{Slug: "general", Name: "General", Content: summary}},
	}
}

// extractJSON pulls the first JSON object out of text, tolerating fences
// and surrounding prose. Mirrors scribe/analyzer.py's _extract_json.
func extractJSON(text string) (map[string]any, error) {
	cleaned := strings.TrimSpace(text)
	// Strip a wrapping ```json ... ``` fence if present.
	if fence := fencePattern.FindStringSubmatch(cleaned); fence != nil {
		cleaned = strings.TrimSpace(fence[1])
	}
	var payload map[string]any
	if err := json.Unmarshal([]byte(cleaned), &payload); err == nil && payload != nil {
		return payload, nil
	}
	// Fall back to the first {...} block in the text.
	start := strings.Index(cleaned, "{")
	end := strings.LastIndex(cleaned, "}")
	if start != -1 && end > start {
		if err := json.Unmarshal([]byte(cleaned[start:end+1]), &payload); err != nil {
			return nil, fmt.Errorf("no parseable JSON object in analyzer response: %w", err)
		}
		return payload, nil
	}
	return nil, fmt.Errorf("no JSON object found in analyzer response")
}

// parseResponse converts the decoded JSON payload into an AnalysisResult.
// Mirrors scribe/analyzer.py's _parse_response.
func parseResponse(payload map[string]any) (*types.AnalysisResult, error) {
	meeting, ok := payload["meeting"].(map[string]any)
	if !ok {
		return nil, fmt.Errorf("missing 'meeting' object in LeMUR response")
	}

	var actionItems []types.ActionItem
	if items, ok := meeting["action_items"].([]any); ok {
		for _, item := range items {
			if obj, ok := item.(map[string]any); ok {
				owner := "unassigned"
				if rawOwner, exists := obj["owner"]; exists {
					owner = pyStr(rawOwner)
				}
				task := ""
				if rawTask, exists := obj["task"]; exists {
					task = pyStr(rawTask)
				}
				actionItems = append(actionItems, types.ActionItem{Owner: owner, Task: task})
			} else if isTruthy(item) {
				actionItems = append(actionItems, types.ActionItem{Owner: "unassigned", Task: pyStr(item)})
			}
		}
	}

	var topics []types.Topic
	if list, ok := payload["topics"].([]any); ok {
		for _, item := range list {
			obj, ok := item.(map[string]any)
			if !ok || !isTruthy(obj["slug"]) {
				continue
			}
			slug := pyStr(obj["slug"])
			name := slug
			if isTruthy(obj["name"]) {
				name = pyStr(obj["name"])
			}
			content := ""
			if isTruthy(obj["content"]) {
				content = pyStr(obj["content"])
			}
			topics = append(topics, types.Topic{Slug: slug, Name: name, Content: content})
		}
	}

	title := "Untitled meeting"
	if isTruthy(meeting["title"]) {
		title = pyStr(meeting["title"])
	}
	summary := ""
	if isTruthy(meeting["summary"]) {
		summary = pyStr(meeting["summary"])
	}

	return &types.AnalysisResult{
		Title:       title,
		Summary:     summary,
		KeyPoints:   asStrList(meeting["key_points"]),
		Decisions:   asStrList(meeting["decisions"]),
		ActionItems: actionItems,
		Topics:      topics,
	}, nil
}

// asStrList returns the items of value as strings, or nil when value is
// not a JSON array. Mirrors _as_str_list.
func asStrList(value any) []string {
	list, ok := value.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(list))
	for _, item := range list {
		out = append(out, pyStr(item))
	}
	return out
}

// isTruthy mirrors Python truthiness for decoded JSON values.
func isTruthy(value any) bool {
	switch v := value.(type) {
	case nil:
		return false
	case bool:
		return v
	case string:
		return v != ""
	case float64:
		return v != 0
	case []any:
		return len(v) > 0
	case map[string]any:
		return len(v) > 0
	default:
		return true
	}
}

// pyStr renders a decoded JSON scalar the way Python's str() would.
func pyStr(value any) string {
	switch v := value.(type) {
	case nil:
		return "None"
	case string:
		return v
	case bool:
		if v {
			return "True"
		}
		return "False"
	default:
		return fmt.Sprintf("%v", v)
	}
}
